File: rest-common/src/main/java/org/chronopolis/rest/entities/UpdatableEntity.java
package org.chronopolis.rest.entities;

import org.chronopolis.rest.ZonedDateTimeConverter;
import org.chronopolis.rest.listener.UpdatableEntityListener;

import javax.persistence.Convert;
import javax.persistence.EntityListeners;
import javax.persistence.MappedSuperclass;
import java.time.ZonedDateTime;

/**
 * Entity which can be updated, and has the latest changes tracked by timestamp
 *
 * Created by shake on 3/21/16.
 */
@MappedSuperclass
@EntityListeners(UpdatableEntityListener.class)
public class UpdatableEntity extends PersistableEntity {

    @Convert(converter = ZonedDateTimeConverter.class)
    private ZonedDateTime createdAt;

    @Convert(converter = ZonedDateTimeConverter.class)
    private ZonedDateTime updatedAt;


    public ZonedDateTime getCreatedAt() {
        return createdAt;
    }

    public UpdatableEntity setCreatedAt(ZonedDateTime createdAt) {
        this.createdAt = createdAt;
        return this;
    }

    public ZonedDateTime getUpdatedAt() {
        return updatedAt;
    }

    public UpdatableEntity setUpdatedAt(ZonedDateTime updatedAt) {
        this.updatedAt = updatedAt;
        return this;
    }


}

File: rest-entities/src/main/kotlin/org/chronopolis/rest/entities/UpdatableEntity.kt
package org.chronopolis.rest.entities

import org.chronopolis.rest.entities.converters.ZonedDateTimeConverter
import java.time.ZonedDateTime
import javax.persistence.Convert
import javax.persistence.MappedSuperclass
import javax.persistence.PreUpdate

@MappedSuperclass
open class UpdatableEntity(
        @Convert(converter = ZonedDateTimeConverter::class)
        var createdAt: ZonedDateTime = ZonedDateTime.now(),

        @Convert(converter = ZonedDateTimeConverter::class)
        var updatedAt: ZonedDateTime = ZonedDateTime.now()
) : PersistableEntity() {

    @PreUpdate
    protected fun onUpdate() {
        updatedAt = ZonedDateTime.now()
    }
}

